purchases/src/test/java/com/revenuecat/purchases/google/usecase/QueryProductDetailsUseCaseTest.kt

```diff
@@ -0,0 +1,621 @@
+    private var capturedConsumeParams = slot<ConsumeParams>()
+
+    private lateinit var wrapper: BillingWrapper
+
+    private lateinit var mockDetailsList: List<ProductDetails>
+
+    private var storeProducts: List<StoreProduct>? = null
+
+    private val billingClientOKResult = BillingClient.BillingResponseCode.OK.buildResult()
+    private var mockActivity = mockk<Activity>()
+
+    @Before
+    fun setup() {
+        storeProducts = null
+        purchasesUpdatedListener = null
+        billingClientStateListener = null
+
+        mockRunnables()
+        mockDiagnosticsTracker()
+        every { mockDateProvider.now } returns Date(1676379370000) // Tuesday, February 14, 2023 12:56:10 PM GMT
+
+        val listenerSlot = slot<PurchasesUpdatedListener>()
+        every {
+            mockClientFactory.buildClient(capture(listenerSlot))
+        } answers {
+            purchasesUpdatedListener = listenerSlot.captured
+            mockClient
+        }
+
+        val billingClientStateListenerSlot = slot<BillingClientStateListener>()
+        every {
+            mockClient.startConnection(capture(billingClientStateListenerSlot))
+        } answers {
+            billingClientStateListener = billingClientStateListenerSlot.captured
+        }
+
+        every {
+            mockClient.endConnection()
+        } just Runs
+
+        every {
+            mockClient.acknowledgePurchase(
+                capture(capturedAcknowledgePurchaseParams),
+                capture(capturedAcknowledgeResponseListener)
+            )
+        } just Runs
+
+        mockConsumeAsync(billingClientOKResult)
+
+        every {
+            mockClient.isReady
+        } returns false andThen true
+
+        val featureSlot = slot<String>()
+        every {
+            mockClient.isFeatureSupported(capture(featureSlot))
+        } returns billingClientOKResult
+
+        mockDetailsList = listOf(mockProductDetails())
+
+        wrapper = BillingWrapper(mockClientFactory, handler, mockDeviceCache, mockDiagnosticsTracker, mockDateProvider)
+        wrapper.purchasesUpdatedListener = mockPurchasesListener
+        wrapper.startConnectionOnMainThread()
+        onConnectedCalled = false
+        wrapper.stateListener = object : BillingAbstract.StateListener {
+            override fun onConnected() {
+                onConnectedCalled = true
+            }
+        }
+
+        every {
+            mockActivity.intent
+        } returns Intent()
+    }
+
+    @After
+    fun tearDown() {
+        clearAllMocks()
+    }
+
+    @Test
+    fun whenProductDetailsIsEmptyPassAnEmptyListToTheListener() {
+        mockEmptyProductDetailsResponse()
+
+        val productIDs = setOf("product_a")
+
+        var receivedList: List<StoreProduct>? = null
+        wrapper.queryProductDetailsAsync(
+            ProductType.SUBS,
+            productIDs, {
+                receivedList = it
+            }, {
+                AssertionsForClassTypes.fail("shouldn't be an error")
```

Comment:
Hmm couldn't this be `Assertions.fail`? I have never used `AssertionsForClassTypes` 😅

Reply:
oh yes, what's this!!! Android Studio did something very weird when I moved things here, I've had to remove a lot of these implicit imports